src/react/components/VisualFlowEditor.jsx

```diff
@@ -0,0 +1,164 @@
+    ))}
+    <div className="block-label">Consumer</div>
+    {filterSortBlocks(blocks, 'consumer').map((block) => (
+      <BlockMenuItem key={block.name} block={block} />
+    ))}
+  </div>
+);
+
+function getEngine(model, nodeFactory) {
+  const engine = createEngine();
+  engine.getNodeFactories().registerFactory(nodeFactory);
+  engine.setModel(model);
+
+  return engine;
+}
+
+const VisualFlowEditor = ({ className = '', blocks, model, onNodeSettingsClick }) => {
+  const [nodeFactory] = useState(new NativeBlockFactory(blocks));
+  const [engine] = useState(getEngine(model, nodeFactory));
+
+  useEffect(() => {
+    nodeFactory.updateDefinitions(blocks);
+  }, [blocks]);
+
+  const addBlock = useCallback(
+    (name, position) => {
+      const newNode = nodeFactory.generateModel({
+        name,
+        onSettingsClick: onNodeSettingsClick,
+      });
+      newNode.setPosition(position.x - 30, position.y - 20);
+      model.addNode(newNode);
+      engine.repaintCanvas();
+    },
+    [onNodeSettingsClick],
+  );
+
+  return (
+    <div className={['flow-editor', className].join(' ')}>
+      <EditorSidebar blocks={blocks} />
+      <div
+        className="canvas-container"
+        onDragOver={(e) => e.preventDefault()}
+        onDrop={(e) => {
+          const blockName = e.dataTransfer.getData('block-name');
+          addBlock(blockName, engine.getRelativeMousePoint(e));
+        }}
+      >
+        <CanvasWidget engine={engine} />
+      </div>
+    </div>
+  );
+};
+
+function getNewModel() {
+  return new DiagramModel();
+}
+
+function nodeModelToSource(model) {
+  const seenIds = new Set();
+  const chain = [];
+  const pipelineBlocks = model.getNodes();
+
+  const sources = pipelineBlocks.filter((b) => b.blockType === 'producer');
+  if (sources.length === 0) {
+    throw new Error('No source block found');
+  }
+
+  if (sources.length > 1) {
+    throw new Error('Too many source blocks, pipelines can have only 1 source block');
+  }
+
+  const sourceBlock = sources[0];
+  chain.push(sourceBlock);
+  seenIds.add(sourceBlock.options.id);
+
+  let prevBlock = sourceBlock;
+  let nextBlock;
+  let loopCounter = 0;
+
+  do {
+    const nextLinks = Object.values(prevBlock.outPorts[0].links);
+
+    if (nextLinks.length === 0) {
+      throw new Error('The pipeline does not end with a consumer block');
+    }
+
+    if (nextLinks.length > 1) {
+      throw new Error('Multiple out connections are not supported');
+    }
+
+    nextBlock = nextLinks[0].targetPort.parent;
+
+    if (seenIds.has(nextBlock.options.id)) {
+      throw new Error('Pipelines cannot form a loop');
+    }
+
+    chain.push(nextBlock);
+    seenIds.add(nextBlock.options.id);
+    prevBlock = nextBlock;
+
+    loopCounter += 1;
+    if (loopCounter > 50) {
+      throw new Error('Pipeline too long');
+    }
+  } while (nextBlock.blockType !== 'consumer');
+
+  return chain.map((block) => block.toScript()).join('\n| ');
+}
```

Comment:
Would it make sense to move this method directly into `models/NativeBlockModel.js` ?

Reply:
This has to do with the pipeline as a whole, not with the single block. I'd leave it in the editor itself, as it is the one that defines and controls the connection between nodes